File: testSuite/test-mass-conservation-coldMode.pl
#!/usr/bin/env perl
use strict;
use warnings;
use Cwd;
use lib $ENV{'GALACTICUS_EXEC_PATH'         }."/perl";
use lib $ENV{'GALACTICUS_ANALYSIS_PERL_PATH'}."/perl";
use PDL;
use PDL::NiceSlice;
use Galacticus::HDF5;

# Run a cold-mode hot halo model to test mass conservation.
# Andrew Benson (23-May-2016)

# Run the model.
system("cd ..; scripts/aux/launch.pl testSuite/parameters/test-mass-conservation-coldMode.xml");

# Check for failed models.
system("grep -q -i fatal outputs/test-mass-conservation-coldMode/galacticus_*/galacticus.log");
if ( $? == 0 ) {
    # Failures were found. Output their reports.
    my @failures = split(" ",`grep -l -i fatal outputs/test-mass-conservation-coldMode/galacticus_*/galacticus.log`);
    foreach my $failure ( @failures ) {
	print "FAILED: log from ".$failure.":\n";
	system("cat ".$failure);
    }
} else {
    print "SUCCESS: model run\n";
}

# Extract masses and check they add up to the expected value.
my $galacticus;
$galacticus->{'file' } = "outputs/test-mass-conservation-coldMode/galacticus_0:1/galacticus.hdf5";
$galacticus->{'store'} = 0;
$galacticus->{'tree' } = "all";
&Galacticus::HDF5::Get_Parameters($galacticus    );
&Galacticus::HDF5::Count_Trees   ($galacticus    );
&Galacticus::HDF5::Select_Output ($galacticus,0.0);
&Galacticus::HDF5::Get_Dataset($galacticus,['mergerTreeWeight','blackHoleMass','diskMassStellar','diskMassGas','spheroidMassStellar','spheroidMassGas','hotHaloMass','hotHaloMassCold','hotHaloOutflowedMass','nodeIsIsolated','basicMass','hotHaloUnaccretedMass','mergerTreeIndex']);
my $properties = $galacticus->{'dataSets'  };
my $parameters = $galacticus->{'parameters'};
# Find centrals.
my $centrals   = which($properties->{'nodeIsIsolated'} == 1);
my $satellites = which($properties->{'nodeIsIsolated'} == 0);
# Find masses in satellites
my $massSatellites = pdl zeroes(nelem($centrals));
for(my $i=0;$i<nelem($centrals);++$i) {
    my $inTree = which($properties->{'mergerTreeIndex'}->($satellites) == $properties->{'mergerTreeIndex'}->($centrals)->(($i)));
    $massSatellites->(($i)) .=
	    (
	     +$properties->{'diskMassStellar'      }->($satellites)->($inTree)->sum()
	     +$properties->{'diskMassGas'          }->($satellites)->($inTree)->sum()	
	     +$properties->{'spheroidMassStellar'  }->($satellites)->($inTree)->sum()
	     +$properties->{'spheroidMassGas'      }->($satellites)->($inTree)->sum()	
	     +$properties->{'hotHaloMass'          }->($satellites)->($inTree)->sum()
	     +$properties->{'hotHaloMassCold'      }->($satellites)->($inTree)->sum()
	     +$properties->{'hotHaloOutflowedMass' }->($satellites)->($inTree)->sum()
	     +$properties->{'hotHaloUnaccretedMass'}->($satellites)->($inTree)->sum()
	     +$properties->{'blackHoleMass'        }->($satellites)->($inTree)->sum()
	    )
	    /(
		+$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
		/$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
	    )
	    /$properties->{'basicMass'}->($centrals)->(($i));
}
$properties->{'diskMassStellar'      } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'diskMassGas'          } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'spheroidMassStellar'  } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'spheroidMassGas'      } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'hotHaloMass'          } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'hotHaloMassCold'      } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'hotHaloOutflowedMass' } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'hotHaloUnaccretedMass'} /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
$properties->{'blackHoleMass'        } /= 
    (
     +$parameters->{'cosmologyParametersMethod'}->{'OmegaBaryon'}->{'value'}
     /$parameters->{'cosmologyParametersMethod'}->{'OmegaMatter'}->{'value'}
    )
    *$properties->{'basicMass'};
my $massTotal =
    +$properties    ->{'diskMassStellar'      }->($centrals)
    +$properties    ->{'diskMassGas'          }->($centrals)
    +$properties    ->{'spheroidMassStellar'  }->($centrals)
    +$properties    ->{'spheroidMassGas'      }->($centrals)
    +$properties    ->{'hotHaloMass'          }->($centrals)
    +$properties    ->{'hotHaloMassCold'      }->($centrals)
    +$properties    ->{'hotHaloOutflowedMass' }->($centrals)
    +$properties    ->{'hotHaloUnaccretedMass'}->($centrals)
    +$properties    ->{'blackHoleMass'        }->($centrals)
    +$massSatellites;
# Check that all masses are unity.
if ( any(abs($massTotal-1.0) > 1.0e-4) ) {
    print "FAILED: mass conservation failure -> ".($massTotal-1.0)."\n";
} else {
    print "SUCCESS: mass conservation\n";
}
exit;

File: testSuite/test-mass-conservation-coldMode.py
#!/usr/bin/env python3
import subprocess
import sys
import glob
import h5py
import numpy as np

# Run a cold mode model to test mass conservation.
# Andrew Benson (ported to Python)

# Run the model.
subprocess.run(
    "cd ..; ./scripts/aux/launch.py testSuite/parameters/test-mass-conservation-coldMode.xml --launchMethod local --threadMaximum 1 --ompThreads 4",
    shell=True
)

# Check for failed models.
logFiles = glob.glob("outputs/test-mass-conservation-coldMode/galacticus_*/galacticus.log")
failures = []
for logFile in logFiles:
    result = subprocess.run(f"grep -q -i -e fatal -e aborted {logFile}", shell=True)
    if result.returncode == 0:
        failures.append(logFile)

if failures:
    for failure in failures:
        print(f"FAILED: log from {failure}:")
        with open(failure) as f:
            print(f.read())
else:
    print("SUCCESS: model run")

# Extract masses and check they add up to the expected value.
with h5py.File("outputs/test-mass-conservation-coldMode/galacticus_0:1/galacticus.hdf5", "r") as galacticus:
    cosmologyParams = galacticus["Parameters/cosmologyParameters"]
    OmegaBaryon     = cosmologyParams.attrs["OmegaBaryon"]
    OmegaMatter     = cosmologyParams.attrs["OmegaMatter"]
    baryonFraction  = OmegaBaryon / OmegaMatter

    nodeData = galacticus["Outputs/Output1/nodeData"]
    props    = {}
    for key in ("mergerTreeWeight", "blackHoleMass", "diskMassStellar", "diskMassGas",
                "spheroidMassStellar", "spheroidMassGas",
                "hotHaloMass", "hotHaloMassCold", "hotHaloOutflowedMass",
                "hotHaloUnaccretedMass", "nodeIsIsolated", "basicMass", "mergerTreeIndex"):
        props[key] = nodeData[key][:]

centrals   = np.where(props["nodeIsIsolated"] == 1)[0]
satellites = np.where(props["nodeIsIsolated"] == 0)[0]

massSatellites = np.zeros(len(centrals))
for i, ci in enumerate(centrals):
    treeIdx = props["mergerTreeIndex"][ci]
    inTree  = np.where(props["mergerTreeIndex"][satellites] == treeIdx)[0]
    satMass = (
        props["diskMassStellar"][satellites][inTree].sum()
        + props["diskMassGas"][satellites][inTree].sum()
        + props["spheroidMassStellar"][satellites][inTree].sum()
        + props["spheroidMassGas"][satellites][inTree].sum()
        + props["hotHaloMass"][satellites][inTree].sum()
        + props["hotHaloMassCold"][satellites][inTree].sum()
        + props["hotHaloOutflowedMass"][satellites][inTree].sum()
        + props["hotHaloUnaccretedMass"][satellites][inTree].sum()
        + props["blackHoleMass"][satellites][inTree].sum()
    )
    massSatellites[i] = satMass / baryonFraction / props["basicMass"][ci]

normFactor = baryonFraction * props["basicMass"]
massTotal = (
    props["diskMassStellar"][centrals]         / normFactor[centrals]
    + props["diskMassGas"][centrals]           / normFactor[centrals]
    + props["spheroidMassStellar"][centrals]   / normFactor[centrals]
    + props["spheroidMassGas"][centrals]       / normFactor[centrals]
    + props["hotHaloMass"][centrals]           / normFactor[centrals]
    + props["hotHaloMassCold"][centrals]       / normFactor[centrals]
    + props["hotHaloOutflowedMass"][centrals]  / normFactor[centrals]
    + props["hotHaloUnaccretedMass"][centrals] / normFactor[centrals]
    + props["blackHoleMass"][centrals]         / normFactor[centrals]
    + massSatellites
)

if np.any(np.abs(massTotal - 1.0) > 1.0e-3):
    print(f"FAILED: mass conservation [coldMode] failure -> {massTotal - 1.0}")
else:
    print("SUCCESS: mass conservation")
